Add support for OpenCLIP

sbalandi opened the pull request:
# What does this PR do?

Support of open_clip library for zero-shot-image-classification task is added by this PR. The models in open_clip format could be loaded and converted to OpenVINO IR by optimum-cli, example:
`optimum-cli export openvino  -m [model_name_or_path] --framework pt [output dir] --weight-format fp32`

And via `OVModel` classes:
```
model = OVModelCLIPForZeroShotImageClassification.from_pretrained([model_name_or_path])
image = processor(image).unsqueeze(0)
text = tokenizer(["a diagram", "a dog", "a cat"])
result = model(text, image)
image_features = model (image).image_features
text_features = model (text).text_features
```

or 

```
visual_model = OVModelCLIPVisual.from_pretrained(model_name_or_path)
text_model  = OVModelCLIPText.from_pretrained(model_name_or_path)
image = processor(image).unsqueeze(0)
text = tokenizer(["a diagram", "a dog", "a cat"])
image_features = visual_model(image).image_features
text_features = text_model(text).text_features
```


have checked on several models:
apple/MobileCLIP-B-LT-OpenCLIP
imageomics/bioclip
laion/CLIP-ViT-B-32-laion2B-s34B-b79K

without open_clip_config:
laion/CLIP-convnext_base-laion400M-s13B-b51K

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

sbalandi commented:
@eaidova please, review

eaidova commented on optimum/exporters/openvino/__main__.py:
```diff
@@ -308,21 +316,61 @@ class StoreAttr(object):
+    model = None
+    if library_name == "open_clip":
+        if is_open_clip_available():
+            import open_clip
+            model, _ = open_clip.create_model_from_pretrained(f'hf-hub:{model_name_or_path}', cache_dir=cache_dir, force_custom_text=True)
+            # tokenizer = open_clip.get_tokenizer(f'hf-hub:{model_name_or_path}')
+            # preprocessors = [tokenizer, preprocessor]
+
+            if not getattr(model, 'config', None):
+                model_config = None
+                try:
+                    model_config = AutoConfig.from_pretrained(model_name_or_path)
+                except:
+                    pass
+
+                if model_config is None:
+                    try:
+                        config_path = hf_hub_download(repo_id=model_name_or_path, filename="open_clip_config.json", cache_dir=cache_dir)
+                        cfg = {}
+                        with open(config_path, 'r', encoding='utf-8') as f:
+                            cfg = json.load(f)
+                        model_config = CLIPConfig(text_config_dict=cfg['model_cfg']['vision_cfg'], vision_config_dict=cfg['model_cfg']['text_cfg'])
+                    except:
+                        pass
+
+                if model_config is None:
+                    raise ImportError(
+                        f"Could not load config for model {model_name_or_path}"
+                    )
+
+                setattr(model, 'config', model_config)
+
+            if not getattr(model.config, 'model_type', None):
+                setattr(model.config, 'model_type', 'clip')
+                setattr(model.config, 'export_model_type', 'clip')
+        else:
+            raise ImportError(
+                "Could not import open_clip module, please, install it to download and convert model"
+            )
```
can it be moved in separated function? I think in future if we will integrate it in common optimum part, it will simplifyrefactoring

sbalandi replied:
moved to OpenClipForZeroShotImageClassification

eaidova commented on optimum/exporters/openvino/convert.py:
```diff
@@ -546,8 +540,12 @@ def export_from_model(
-    library_name = TasksManager._infer_library_from_model_or_model_class(model=model)
-    TasksManager.standardize_model_attributes(model)
+    library_name = None
+    if "zero-shot-image-classification" in task and 'clip' in getattr(model.config, 'model_type'):
```
are you

eaidova commented on optimum/exporters/openvino/convert.py:
```diff
@@ -546,8 +540,12 @@ def export_from_model(
-    library_name = TasksManager._infer_library_from_model_or_model_class(model=model)
-    TasksManager.standardize_model_attributes(model)
+    library_name = None
+    if "zero-shot-image-classification" in task and 'clip' in getattr(model.config, 'model_type'):
```
can we have more robust way to understand that it is open_clip? TaskManager checks available files for that, can existence open_clip_config.json be signal that we have open_clip model?

sbalandi replied:
done with ov_infer_library_from_model

eaidova commented:
@sbalandi export part generally looks good.
I have 2 questions:
1. Could you please add tests?
2. Is there is some plans to add inference model classes? I guess case, when model exported as separated image and text models, it required some additional logic for making prediction, right?

openvino-dev-samples commented:
I have a question on this PR
How can we select to export a consolidated model or 2 separate models (text + vision) ?

sbalandi commented:
> @sbalandi export part generally looks good. I have 2 questions:
> 
> 1. Could you please add tests?
> 2. Is there is some plans to add inference model classes? I guess case, when model exported as separated image and text models, it required some additional logic for making prediction, right?

added test to tests/openvino/test_exporters_cli.py
added classes OVModellCLIPText/OVModellCLIPVision

eaidova commented on optimum/intel/openvino/modeling.py:
```diff
@@ -1031,3 +1035,183 @@ def forward(self, **kwargs):
+    """,
+    MODEL_START_DOCSTRING,
+)
+class OVModellCLIPVisual(OVModel):
```
Should we also have class for ZeroShotClassification too that will use both text and vision models?

sbalandi replied:
I have added OVModellCLIPForZeroShotImageClassification with parent PreTrainedModel and text/vision_model which are OVBaseModel. Also I have implemented methods from OVBaseModel , but haven't added attributes such as output_names/input_names. Is it what expected ? Should I add consolidated model and choice what model user will use ?

eaidova commented on tests/openvino/utils_tests.py:
```diff
@@ -137,6 +137,7 @@
     "xglm": "hf-internal-testing/tiny-random-XGLMForCausalLM",
     "xverse": "katuni4ka/tiny-random-xverse",
     "glm4": "katuni4ka/tiny-random-glm4",
+    'open-clip': 'laion/CLIP-ViT-bigG-14-laion2B-39B-b160k'
```
wow, I suppose this model too big for testing, can we use some tiny version?

sbalandi replied:
changed to laion/CLIP-ViT-L-14-laion2B-s32B-b82K )

sbalandi commented on optimum/intel/openvino/modeling.py:
```diff
@@ -1031,3 +1035,183 @@ def forward(self, **kwargs):
+        if local_open_clip_model or (not os.path.isdir(model_id) and model_info(model_id).library_name == "open_clip"):
+            if not is_open_clip_available():
+                raise ImportError(
+                    "To load a timm model, timm needs to be installed. Please install it with `pip install timm`."
```
moved

sbalandi commented:
> How can we select to export a consolidated model or 2 separate models (text + vision) ?

Now we export two separate models only

sbalandi commented on optimum/intel/openvino/modeling.py:
```diff
@@ -1031,3 +1035,183 @@ def forward(self, **kwargs):
+                raise ValueError("You cannot use both `use_auth_token` and `token` arguments at the same time.")
+            token = use_auth_token
+            
+        # Fix the mismatch between timm_config and huggingface_config
```
removed

eaidova commented on optimum/exporters/openvino/__main__.py:
```diff
@@ -54,25 +58,29 @@ def infer_task(
+                )
+            except KeyError as e:
+                raise KeyError(
+                    f"The task could not be automatically inferred. Please provide the argument --task with the relevant task from {', '.join(TasksManager.get_all_tasks())}. Detailed error: {e}"
```
TaskManager return list of tasks without open_clip or not? Should it be explicitly included

sbalandi replied:
TaskManager.get_all_tasks() include zero-shot-image-classification task

eaidova commented on tests/openvino/utils_tests.py:
```diff
@@ -137,6 +137,7 @@
     "xglm": "hf-internal-testing/tiny-random-XGLMForCausalLM",
     "xverse": "katuni4ka/tiny-random-xverse",
     "glm4": "katuni4ka/tiny-random-glm4",
+    'open-clip': 'laion/CLIP-ViT-L-14-laion2B-s32B-b82K'
```
2Gb, still big for testing, can we try this one (it is smallest what I found https://huggingface.co/apple/MobileCLIP-S1-OpenCLIP)?

sbalandi replied:
done

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/optimum-intel/pr_857). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

eaidova commented:
@sbalandi could you please also include package version into meta https://github.com/huggingface/optimum-intel/blob/main/optimum/exporters/openvino/convert.py#L775

sbalandi commented:
> @sbalandi could you please also include package version into IR meta https://github.com/huggingface/optimum-intel/blob/main/optimum/exporters/openvino/convert.py#L775

added

eaidova commented on optimum/intel/openvino/modeling.py:
```diff
@@ -1030,3 +1040,356 @@ def forward(self, **kwargs):
+        visual_model_outputs = self.visual_model._inference({"x": pixel_values})
+
+        return ModelOutput(
+            image_features=visual_model_outputs["image_features"], text_features=text_model_outputs["text_features"]
```
I believe zero shot classification means that you need return not only features, but also probabilities like here https://github.com/huggingface/transformers/blob/main/src/transformers/models/clip/modeling_clip.py#L1334

BTW, can these model classes can also handle transformers clip realization? if only open_clip expected, then it is better rename model class

sbalandi replied:
names are updated, output are updated too according to transformers clip model output

eaidova commented:
@sbalandi please also provide tests for openclip model wrappers in test_modeling.py

eaidova commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -954,3 +962,106 @@ def patch_model_for_export(
+        model_inputs["x"] = inputs["pixel_values"]
+        return model_inputs
+
+    def generate_dummy_inputs(self, framework: str = "pt", **kwargs):
+        return super().generate_dummy_inputs(framework, **kwargs)
```
```suggestion
```

sbalandi replied:
removed

eaidova commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -954,3 +962,106 @@ def patch_model_for_export(
+        return OpenCLIPModelPatcher(self, model, model_kwargs=model_kwargs)
+
+
+@register_in_tasks_manager("clip_vision_model", *["zero-shot-image-classification"], library_name="open_clip")
```
I believe feature_extraction is also possible task name

eaidova commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -954,3 +962,106 @@ def patch_model_for_export(
+        return OpenCLIPModelPatcher(self, model, model_kwargs=model_kwargs)
+
+
+@register_in_tasks_manager("clip_text_model", *["zero-shot-image-classification"], library_name="open_clip")
```
here also if we allo to export separated models feature_extraction can be possible usecase

sbalandi commented:
@IlyasMoutawwakil @echarlaix could you please trigger ci ?

sbalandi commented:
> @sbalandi please also provide tests for openclip model wrappers in test_modeling.py

added

eaidova commented:
@echarlaix  could you please help to trigger precommit? Thanks!

echarlaix commented on tests/openvino/test_modeling.py:
```diff
@@ -1943,3 +1947,154 @@ def test_compare_output_hidden_states(self, model_arch):
+        loaded_model_outputs = loaded_model(tokens, processed_image)
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            loaded_model.save_pretrained("/home/labuser/work/notebook/optimum-intel/clip_ov_42")
```
```suggestion
```

sbalandi replied:
removed

echarlaix commented on optimum/exporters/openvino/utils.py:
```diff
@@ -141,3 +149,37 @@ def clear_class_registry():
     torch.jit._state._clear_class_state()
+
+
+def ov_infer_library_from_model_name_or_path(
```
```suggestion
def _infer_library_from_model_name_or_path(
```

sbalandi replied:
renamed

echarlaix commented on optimum/exporters/openvino/utils.py:
```diff
@@ -141,3 +149,37 @@ def clear_class_registry():
+    return library_name
+
+
+def ov_infer_library_from_model_or_model_class(
```
```suggestion
def _infer_library_from_model_or_model_class(
```

sbalandi replied:
renamed

echarlaix commented on optimum/intel/utils/import_utils.py:
```diff
@@ -113,6 +113,17 @@
+_open_clip_version = "N/A"
+if _open_clip_available:
+    try:
+        import open_clip
+
+        _open_clip_version = open_clip.__version__
```
why not 

```suggestion
        _open_clip_version = importlib_metadata.version("open_clip_torch")
```

sbalandi replied:
open_clip.__version__  is not work for open_clip

echarlaix replied:
thanks for the update, `_open_clip_version = importlib_metadata.version("open_clip_torch")` works for me, could you extend on when this won't work as expected ?

echarlaix commented on optimum/intel/openvino/modeling.py:
```diff
@@ -1030,3 +1044,459 @@ def forward(self, **kwargs):
+    """,
+    MODEL_START_DOCSTRING,
+)
+class OVModelOpenCLIPVisual(OVModel):
```
can you move everything in a separate file like modeling_open_clip.py

sbalandi replied:
moved

echarlaix commented on optimum/exporters/openvino/convert.py:
```diff
@@ -662,7 +657,38 @@ def export_from_model(
 
-    if library_name != "diffusers":
+    if library_name == "open_clip":
+        models_and_export_configs = {}
+
+        visual_model = model.visual
+        setattr(visual_model, "config", model.config.vision_config)
+        vision_cfg = OpenCLIPVisualOpenVINOConfig(
+            model.config.vision_config, "zero-shot-image-classification", preprocessors
+        )
+        models_and_export_configs["vision"] = (visual_model, vision_cfg)
+
+        text_model = model.text
+        setattr(text_model, "config", model.config.text_config)
+        text_cfg = OpenCLIPTextOpenVINOConfig(
+            model.config.text_config, "zero-shot-image-classification", preprocessors
+        )
+        models_and_export_configs["text"] = (text_model, text_cfg)
+
```
could this be moved in a separate function as done for other models  with `_get_submodels_and_export_configs`

sbalandi replied:
moved

echarlaix commented on optimum/exporters/openvino/convert.py:
```diff
@@ -662,7 +657,38 @@ def export_from_model(
+        if hasattr(model.config, "save_pretrained"):
+            model.config.save_pretrained(output)
+
+        try:
+            hf_hub_download(repo_id=model_name_or_path, filename="open_clip_config.json", cache_dir=output)
+        except:
+            pass
```
why is this needed ?

sbalandi replied:
removed

sbalandi commented:
@echarlaix could you please help to trigger precommit? Thanks!

sbalandi commented:
@echarlaix @IlyasMoutawwakil @AlexKoff88  could you please help to trigger precommit? Thanks!

echarlaix commented on optimum/exporters/openvino/__main__.py:
```diff
@@ -305,21 +321,25 @@ class StoreAttr(object):
-        library_name=library_name,
-        **loading_kwargs,
-    )
+    model = None
```
```suggestion
```

sbalandi replied:
removed

echarlaix commented on optimum/intel/openvino/utils.py:
```diff
@@ -178,6 +181,23 @@ def _is_timm_ov_dir(model_dir):
     return False
 
 
+def _is_open_clip_ov_dir(model_dir, model_name):
+    config_file = None
+    has_xml = False
+    has_bin = False
+    if os.path.isdir(model_dir):
+        for filename in glob(os.path.join(model_dir, "*")):
+            if filename.endswith(f"{model_name}.xml"):
+                has_xml = True
+            if filename.endswith(f"{model_name}.bin"):
+                has_bin = True
+            if filename.endswith("config.json"):
+                config_file = filename
+    if config_file and has_xml and has_bin:
+        return True
+    return False
+
+
```
I don't see `_is_open_clip_ov_dir` used anywhere can we remove ? 

```suggestion
```

sbalandi replied:
removed

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+                )
+            model_id, revision = model_id.split("@")
+
+        config_path = config if isinstance(config, (str, os.PathLike)) else model_id
```
if this is the only difference then I think we don't need to rewrite `from_pretrained` but only load the config before calling super().from_pretrained(..) wdyt ?

sbalandi replied:
OptimizedModel.from_pretrained use also `TasksManager.infer_library_from_model` , which will fails for open_clip

echarlaix replied:
ok so this should be added in optimum but let's keep it in optimum-intel for now, can you add a comment to specify that this should be removed when added to optimum (when compatible with TasksManager.infer_library_from_model)

echarlaix replied:
thanks!

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+            **kwargs,
+        )
+
+    @classmethod
+    def _from_pretrained(
+        cls,
+        model_id: Union[str, Path],
+        config: PretrainedConfig,
+        use_auth_token: Optional[Union[bool, str]] = None,
+        token: Optional[Union[bool, str]] = None,
+        revision: Optional[str] = None,
+        force_download: bool = False,
+        cache_dir: str = HUGGINGFACE_HUB_CACHE,
+        file_name: Optional[str] = "openvino_model_text.xml",
+        subfolder: str = "",
+        from_onnx: bool = False,
+        local_files_only: bool = False,
+        load_in_8bit: bool = False,
+        quantization_config: Union[OVWeightQuantizationConfig, Dict] = None,
+        **kwargs,
+    ):
+        return super()._from_pretrained(
+            model_id=model_id,
+            config=config,
+            use_auth_token=use_auth_token,
+            token=token,
+            revision=revision,
+            force_download=force_download,
+            cache_dir=cache_dir,
+            file_name=file_name,
+            subfolder=subfolder,
+            from_onnx=from_onnx,
+            local_files_only=local_files_only,
+            load_in_8bit=load_in_8bit,
+            quantization_config=quantization_config,
+            **kwargs,
+        )
```
why not remove 
```suggestion
```

sbalandi replied:
I added it to setup file name for text/vision model `file_name: Optional[str] = "openvino_model_text.xml",` , without this we will look for `openvino_model.xml`

echarlaix replied:
what about using `cls.OV_XML_FILE_NAME` in https://github.com/huggingface/optimum-intel/blob/ba74307dfc7f69c6281e811e02ee0203a6fb8355/optimum/intel/openvino/modeling_base.py#L325 directly, maybe renaming to something like `cls._ov_xml_file` or `cls._default_xml_file`

echarlaix commented on optimum/intel/utils/modeling_utils.py:
```diff
@@ -226,3 +235,199 @@ def bind_cores_for_best_perf():
+    return model_files
+
+
+class OpenClipForZeroShotImageClassification(PreTrainedModel):
```
why do we need OpenClipForZeroShotImageClassification / cannot load the model with  `open_clip.create_model_and_transforms` directly ?

sbalandi replied:
to be able to load models from a local path and load models without open_clip_config.json on the hugging face hub, for example laion/CLIP-convnext_base-laion400M-s13B-b51K

eaidova commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -74,6 +78,10 @@
 
 
 def init_model_configs():
+    library_to_task_map = getattr(TasksManager, "_LIBRARY_TO_SUPPORTED_MODEL_TYPES")
+    library_to_task_map["open_clip"] = {}
+    setattr(TasksManager, "_LIBRARY_TO_SUPPORTED_MODEL_TYPES", library_to_task_map)
```
```suggestion
TasksManager._LIBRARY_TO_SUPPORTED_MODEL_TYPES["open_clip"] = {}
```

sbalandi replied:
done

eaidova commented on optimum/exporters/openvino/utils.py:
```diff
@@ -25,6 +27,8 @@
     import torch.nn as nn
     from transformers.modeling_utils import PreTrainedModel
 
+    # from transformers.modeling_tf_utils import TFPreTrainedModel
```
does it should be commented?

sbalandi replied:
removed

eaidova commented on tests/openvino/utils_tests.py:
```diff
@@ -141,6 +141,8 @@
     "xverse": "katuni4ka/tiny-random-xverse",
     "glm4": "katuni4ka/tiny-random-glm4",
+    "open-clip": "hf-internal-testing/tiny-open-clip-model",
+    "open-clip-ir": "zofinka/tiny-open-clip-model",
```
for consistency with others (ir is very general name)
```suggestion
    "open-clip-ov": "zofinka/tiny-open-clip-model",
```

sbalandi replied:
renamed

eaidova commented:
@echarlaix could you please take a look one more time? Thanks!

echarlaix commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -74,6 +76,8 @@
 
 
 def init_model_configs():
+    TasksManager._LIBRARY_TO_SUPPORTED_MODEL_TYPES["open_clip"] = {}
```
shouldn't overwrite if exist (could be the case for other export than openvino)
```suggestion
    if "open_clip" not in TasksManager._LIBRARY_TO_SUPPORTED_MODEL_TYPES:
        TasksManager._LIBRARY_TO_SUPPORTED_MODEL_TYPES["open_clip"] = {}
```

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+        config_name_or_path: Union[str, os.PathLike],
+        revision: Optional[str] = None,
+        cache_dir: str = HUGGINGFACE_HUB_CACHE,
+        use_auth_token: Optional[Union[bool, str]] = None,
+        token: Optional[Union[bool, str]] = None,
+        force_download: bool = False,
+        subfolder: str = "",
+        trust_remote_code: bool = False,
+        local_files_only: bool = False,
+    ) -> PretrainedConfig:
+        if use_auth_token is not None:
+            warnings.warn(
+                "The `use_auth_token` argument is deprecated and will be removed soon. Please use the `token` argument instead.",
+                FutureWarning,
+            )
+            if token is not None:
+                raise ValueError("You cannot use both `use_auth_token` and `token` arguments at the same time.")
+            token = use_auth_token
```
shouldn't be needed (already present in from_pretrained)
```suggestion
        token: Optional[Union[bool, str]] = None,
        force_download: bool = False,
        subfolder: str = "",
        trust_remote_code: bool = False,
        local_files_only: bool = False,
    ) -> PretrainedConfig:
```

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+        main_export(
+            model_name_or_path=model_id,
+            output=save_dir_path,
+            task=task or "zero-shot-image-classification",
```
can be set as `export_feature` https://github.com/huggingface/optimum-intel/blob/ba74307dfc7f69c6281e811e02ee0203a6fb8355/optimum/intel/openvino/modeling.py#L309

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+            force_download=force_download,
+            trust_remote_code=trust_remote_code,
+            ov_config=ov_config,
+            library_name="open_clip",
```
can be set as `_library_name` https://github.com/huggingface/optimum-intel/blob/ba74307dfc7f69c6281e811e02ee0203a6fb8355/optimum/intel/openvino/modeling_base.py#L63

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+        outputs = self._inference(inputs)
+        return ModelOutput(text_features=torch.from_numpy(outputs["text_features"]))
+
+    def _save_pretrained(self, save_directory: Union[str, Path]):
+        dst_path = os.path.join(save_directory, self.OV_XML_FILE_NAME)
+        openvino.save_model(self.model, dst_path, compress_to_fp16=False)
+        generation_config = getattr(self, "generation_config", None)
+        if generation_config is not None:
+            try:
+                generation_config.save_pretrained(save_directory)
+            except Exception as exception:
+                logger.warning(
+                    f"The generation config will not be saved, saving failed with following error:\n{exception}"
+                )
+
+        self._save_openvino_config(save_directory)
```
instead of overwriting `_save_pretrained` we add this change here https://github.com/huggingface/optimum-intel/blob/ba74307dfc7f69c6281e811e02ee0203a6fb8355/optimum/intel/openvino/modeling_base.py#L259 instead

echarlaix replied:
I'm suggesting to add this change in `OVBaseModel._save_pretrained` directly (with a `default_file_name` attribute that would be `"openvino_model_text.xml"` for `OVModelOpenCLIPText`, `"openvino_model_vision.xml"` for `OVModelOpenCLIPVisual`  and `"openvino_model.xml"` for the rest) wdyt ?

sbalandi replied:
I have added `xml_model_name `as attribute to `OVBaseModel` , used it in `_save_pretrained` and overwrited it in `OVModelOpenCLIPText`/`OVModelOpenCLIPVisual` Could you please check if I understand you right ?

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+    """,
+    MODEL_START_DOCSTRING,
+)
+class OVModelOpenCLIPForZeroShotImageClassification(torch.nn.Module):
```
Why inheriting from `torch.nn.Module` ? This shouldn't be needed

sbalandi commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+                )
+                logger.info(
+                    f"config.json not found in the specified subfolder {subfolder}. Using the top level config.json."
+                )
```
yes, removed

sbalandi commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+            elif cls.CONFIG_NAME in all_files:
+                config_path = hf_hub_download(
+                    repo_id=config_name_or_path,
+                    filename=cls.CONFIG_NAME,
```
for example if we load converted .xml models from hf_hub

sbalandi commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,733 @@
+                    local_files_only=local_files_only,
+                )
+            else:
+                open_clip_config = OpenClipForZeroShotImageClassification.find_config_by_hub_url(config_name_or_path)
```
updated

sbalandi commented:
@echarlaix could you please take a look ?

sbalandi commented:
@echarlaix @IlyasMoutawwakil could you please trigger precommit ?

sbalandi commented:
@echarlaix could you please take a look ? I have updated the code according to the comments.

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+                The image width.
+        """
+        self.text_model.reshape(batch_size=batch_size, sequence_length=sequence_length, height=height, width=width)
+        self.visual_model.compile(batch_size=batch_size, sequence_length=sequence_length, height=height, width=width)
```
```suggestion
        self.visual_model.reshape(batch_size=batch_size, sequence_length=sequence_length, height=height, width=width)
```

echarlaix replied:
could you add a test to verify this works as expected ?

echarlaix replied:
also we should add test for .to() and .compile()

sbalandi replied:
added

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+            ov_config = None
+        else:
+            ov_config = OVConfig(dtype="fp32")
+
+        def fn_get_submodels(model):
+            return {"model_vision": model.visual}
+
+        custom_export_configs = {
+            "model_vision": {},
+        }
```
question : couldn't this be inferred by `_get_open_clip_submodels_fn_and_export_configs` directly (depending on a given task) ?

echarlaix commented on optimum/exporters/openvino/model_configs.py:
```diff
@@ -1018,3 +1023,118 @@ def patch_model_for_export(
+
+
+@register_in_tasks_manager(
+    "clip-text-model", *["zero-shot-image-classification", "feature-extraction"], library_name="open_clip"
```
The task "zero-shot-image-classification" shouldn't be added here no ? Instead can be fixed in directly `_get_open_clip_submodels_fn_and_export_configs`

```suggestion
    "clip-text-model", *["feature-extraction"], library_name="open_clip"
```

echarlaix commented on optimum/intel/utils/modeling_utils.py:
```diff
@@ -229,4 +238,209 @@ def bind_cores_for_best_perf():
+
+                if not getattr(model, "config", None):
+                    config_path = hf_hub_download(
+                        repo_id=model_name_or_path, filename="open_clip_config.json", cache_dir=cache_dir
```
we could also give argument like revision, subfolder, token, force_download, trust_remote_code, local_files_only

echarlaix commented on optimum/intel/utils/modeling_utils.py:
```diff
@@ -229,4 +238,209 @@ def bind_cores_for_best_perf():
+                model_file_name = None
+                if file_exists(model_name_or_path, "open_clip_pytorch_model.bin"):
+                    model_file_name = "open_clip_pytorch_model.bin"
+                elif file_exists(model_name_or_path, "pytorch_model.bin"):
+                    model_file_name = "pytorch_model.bin"
```
in which case would this happen ?

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+            )
+            export = from_transformers
+
+        if len(model_id.split("@")) == 2:
+            if revision is not None:
+                logger.warning(
+                    f"The argument `revision` was set to {revision} but will be ignored for {model_id.split('@')[1]}"
+                )
+            model_id, revision = model_id.split("@")
```
this will be deprecated in the next optimum release so no need to add it 
```suggestion
```

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+        trust_remote_code: bool = False,
+        **kwargs,
+    ):
+        if use_auth_token is not None:
+            warnings.warn(
+                "The `use_auth_token` argument is deprecated and will be removed soon. Please use the `token` argument instead.",
+                FutureWarning,
+            )
+            if token is not None:
+                raise ValueError("You cannot use both `use_auth_token` and `token` arguments at the same time.")
+            token = use_auth_token
```
same this wan never supported in the first place so would not even allow it
```suggestion
```

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+        if isinstance(model_id, Path):
+            model_id = model_id.as_posix()
+
+        from_transformers = kwargs.pop("from_transformers", None)
+        if from_transformers is not None:
+            logger.warning(
+                "The argument `from_transformers` is deprecated, and will be removed in optimum 2.0.  Use `export` instead"
+            )
+            export = from_transformers
```
same this wan never supported in the first place so would not even allow it
```suggestion
```

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+        trust_remote_code: bool = False,
+        **kwargs,
+    ):
+        if use_auth_token is not None:
```
also what about  automatic conversion if the model wasn't already converted https://github.com/huggingface/optimum-intel/blob/b760a1ddcb663e94b69fbe231e088028262824ac/optimum/intel/openvino/modeling_base.py#L405

sbalandi replied:
added

echarlaix commented on optimum/intel/openvino/modeling_open_clip.py:
```diff
@@ -0,0 +1,680 @@
+        task: Optional[str] = None,
+        trust_remote_code: bool = False,
+        load_in_8bit: Optional[bool] = None,
+        quantization_config: Union[OVWeightQuantizationConfig, Dict] = None,
```
could we add a test for quantization as well  ?

sbalandi replied:
added case to test_ovmodel_load_with_compressed_weights and test_ovmodel_load_with_uncompressed_weights

sbalandi commented:
@echarlaix could you please take a look ? I have updated the code according to the comments

echarlaix commented:
LGTM, thanks a lot for iterating on this PR @sbalandi

eaidova commented on optimum/exporters/openvino/model_patcher.py:
```diff
@@ -2469,6 +2471,45 @@ def patched_forward(*args, **kwargs):
+    elif pool_type == "last":
+        pooled, tokens = x[:, -1], x[:, :-1]
+    elif pool_type == "argmax":
+        text = text.to(dtype=torch.int32)  # ONNX Runtime is unable to run argmax with int64 input, hence this cast.
```
should be not needed

sbalandi replied:
removed

sbalandi commented:
> LGTM, thanks a lot for iterating on this PR @sbalandi

thanks a lot for careful review !
